Feat(eos_designs): Add support to bind IPsec connection to source int…

ashenoy-arista opened the pull request:
## Change Summary

Add support to eos_designs to bind IPsec connection to source interface 

## Related Issue(s)

Fixes #5189 

## Component(s) name

`arista.avd.eos_designs`

## Proposed changes
In CV-Pathfinder solution, when there are multiple ECMP paths available on wan-router to reach destination,
add setting for ipsec to allow binding of ipsec connection to the source interface and ignore available ECMP paths.

Proposal is to add "ipsec_settings" key as dictionary under root.
It would include bool setting "bind_connection_to_interface" (default is false)

- One could add this to yml file under `group_vars` or `host_vars` directory to configure multiple devices vs single device respectively.
- The setting would apply for wan_routers (regardless of their role as client or server) i.e. edge/transit/pathfinder/rr 
- It would be relevant when `wan_mode` is set as  `autovpn` or `cv-pathfinder` both when ipsec is used.
- It would also apply in case of internet exit scenarios that configure IPsec tunnels to reach destination.

<!--- Describe data model implemented for new features -->

## How to test
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
Updated existing unit test input yml under `eos_designs_unit_tests` to validate generated structured config and EOS cli cfg.

## Checklist

### User Checklist

<!-- Add your own checklist using MD syntax and by replacing N/A -->
- N/A

### Repository Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ x ] My code has been rebased from devel before I start
- [ x ] I have read the [**CONTRIBUTING**](https://avd.arista.com/stable/docs/contribution/overview.html) document.
- [ x ] My change requires a change to the documentation and documentation have been updated accordingly.
- [ x ] I have updated [molecule CI](https://github.com/aristanetworks/avd/tree/devel/ansible_collections/arista/avd/molecule) testing accordingly. (check the box if not applicable)

ashenoy-arista committed:
Feat(eos_designs): Add support to bind IPsec connection to source interface

ClausHolbechArista commented on ansible_collections/arista/avd/molecule/eos_designs_unit_tests/inventory/host_vars/cv-pathfinder-edge-custom-default-policy.yml:
```diff
@@ -49,6 +49,8 @@ wan_ipsec_profiles:
     shared_key: ABCDEF1234567890
   data_plane:
     shared_key: ABCDEF1234567890666
+  bind_connection_to_interface:
```
This is not part of the profile, but a common key, so I think we should move this to "root" and call it `ipsec_bind_to_interface: <bool>` and have it default false for now. Also add a comment with "TODO: 6.0.0" in the schema fragment to _consider_ to change the default to true. IMO this behavior would be the expectation when you set a source IP.

ClausHolbechArista replied:
alternatively some `ipsec_settings` key so we can add other things under that.

ashenoy-arista replied:
Adding `ipsec_settings` key under root.
This will allow us to extend this to other parameters in future

ClausHolbechArista commented on python-avd/pyavd/_eos_designs/structured_config/overlay/ip_security.py:
```diff
@@ -118,3 +120,14 @@ def _set_key_controller(self: AvdStructuredConfigOverlayProtocol, profile_name:
+        """
+        Set binding of ipsec connection to the source interface.
+
+        This is applicable if the device is not a RR or pathfinder.
```
IMO we should just set the knob if the user gave it. I think it would also be used for the control-plane tunnels, so also relevant on the pathfinders.

ashenoy-arista replied:
Ack, applying to ALL wan-routers regardless of their role as client or server.

ashenoy-arista committed:
Merge branch 'devel' into eosDesignsBindIPSecConnSrcIntf

ClausHolbechArista committed:
Merge branch 'devel' into eosDesignsBindIPSecConnSrcIntf

ClausHolbechArista commented on python-avd/pyavd/_eos_designs/schema/schema_fragments/ipsec_settings.schema.yml:
```diff
@@ -0,0 +1,22 @@
+type: dict
+keys:
+  ipsec_settings:
+    documentation_options:
+      table: ipsec-settings
```
For now let's keep it with other wan settings. The generated tables will now show up in the docs unless you add an include statement in the "input-variables.md" file.

```suggestion
    documentation_options:
      table: wan-settings
```

ClausHolbechArista commented on python-avd/pyavd/_eos_designs/structured_config/overlay/ip_security.py:
```diff
@@ -118,3 +120,8 @@ def _set_key_controller(self: AvdStructuredConfigOverlayProtocol, profile_name:
         self.structured_config.ip_security.key_controller.profile = profile_name
+
+    def _set_ipsec_parameters(self: AvdStructuredConfigOverlayProtocol) -> None:
+        """Set parameters relevant to all ipsec connections."""
```
```suggestion
        """
        Set parameters relevant to all ipsec connections.
        
        Note: Duplicated in network_services/ip_security.py.
        """
```

ClausHolbechArista commented on python-avd/pyavd/_eos_designs/structured_config/network_services/ip_security.py:
```diff
@@ -62,3 +65,10 @@ def ip_security(self: AvdStructuredConfigNetworkServicesProtocol) -> None:
+            self._set_ipsec_parameters()
+
+    def _set_ipsec_parameters(self: AvdStructuredConfigNetworkServicesProtocol) -> None:
+        """Set any parameters relevant to all ipsec connections."""
```
```suggestion
        """
        Set any parameters relevant to all ipsec connections.

        Note: Duplicated in overlay/ip_security.py.
        """
```

ClausHolbechArista committed:
Update python-avd/pyavd/_eos_designs/structured_config/network_services/ip_security.py

ClausHolbechArista committed:
Update python-avd/pyavd/_eos_designs/structured_config/overlay/ip_security.py

ClausHolbechArista committed:
Update python-avd/pyavd/_eos_designs/schema/schema_fragments/ipsec_settings.schema.yml

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ashenoy-arista committed:
Merge branch 'devel' into eosDesignsBindIPSecConnSrcIntf

gmuloc committed:
Merge branch 'devel' into eosDesignsBindIPSecConnSrcIntf

gmuloc commented:
need to add a test to exercise the code path in network services

gmuloc committed:
Add unit test to account for scenario wherein wan cv-pathfinder Dps tunnels are ipsec disabled,
while configuring internet exit policy that uses IPsec based tunnel.
This exercises the code path in network services.

ashenoy-arista committed:
Merge branch 'devel' into eosDesignsBindIPSecConnSrcIntf

gmuloc committed:
Merge branch 'devel' into eosDesignsBindIPSecConnSrcIntf

gmuloc committed:
Merge branch 'devel' into eosDesignsBindIPSecConnSrcIntf

Vibhu-gslab commented on python-avd/pyavd/_eos_designs/structured_config/overlay/ip_security.py:
```diff
@@ -118,3 +120,12 @@ def _set_key_controller(self: AvdStructuredConfigOverlayProtocol, profile_name:
+
+        Note: Duplicated in network_services/ip_security.py.
+        """
+        if self.inputs.ipsec_settings.bind_connection_to_interface:
+            self.structured_config.ip_security.connection_tx_interface_match_source_ip = True
```
Since this is single line change may be we can move it in main function (ip_security).

Vibhu-gslab replied:
also the one in network_services/ip_security.py.

gmuloc replied:
agreed lets simplify for now

Vibhu-gslab commented on python-avd/pyavd/_eos_designs/schema/schema_fragments/ipsec_settings.schema.yml:
```diff
@@ -0,0 +1,22 @@
+    keys:
+      bind_connection_to_interface:
+        type: bool
+        # TODO: 6.0.0 - Consider changing default to true
+        default: false
```
Can we set the default to true now, given that this is a new data model?
@gmuloc @ClausHolbechArista

ClausHolbechArista replied:
No since it will generate new config without the user changing any inputs.

gmuloc commented on python-avd/pyavd/_eos_designs/structured_config/overlay/ip_security.py:
```diff
@@ -118,3 +120,12 @@ def _set_key_controller(self: AvdStructuredConfigOverlayProtocol, profile_name:
         if self.shared_utils.is_wan_server:
             return
         self.structured_config.ip_security.key_controller.profile = profile_name
+
+    def _set_ipsec_parameters(self: AvdStructuredConfigOverlayProtocol) -> None:
+        """
+        Set parameters relevant to all ipsec connections.
+
+        Note: Duplicated in network_services/ip_security.py.
+        """
+        if self.inputs.ipsec_settings.bind_connection_to_interface:
+            self.structured_config.ip_security.connection_tx_interface_match_source_ip = True
```
```suggestion
```

gmuloc commented on python-avd/pyavd/_eos_designs/structured_config/overlay/ip_security.py:
```diff
@@ -43,6 +43,8 @@ def ip_security(self: AvdStructuredConfigOverlayProtocol) -> None:
             self._set_data_plane()
         self._set_control_plane()
+        # settings applicable to all ipsec connections
+        self._set_ipsec_parameters()
```
```suggestion
        if self.inputs.ipsec_settings.bind_connection_to_interface:
            self.structured_config.ip_security.connection_tx_interface_match_source_ip = True
```

gmuloc commented on python-avd/pyavd/_eos_designs/structured_config/network_services/ip_security.py:
```diff
@@ -62,3 +65,14 @@ def ip_security(self: AvdStructuredConfigNetworkServicesProtocol) -> None:
                 ),
                 connection="start",
             )
+        if is_ipsec_policy_in_use:
+            self._set_ipsec_parameters()
+
+    def _set_ipsec_parameters(self: AvdStructuredConfigNetworkServicesProtocol) -> None:
+        """
+        Set any parameters relevant to all ipsec connections.
+
+        Note: Duplicated in overlay/ip_security.py.
+        """
+        if self.inputs.ipsec_settings.bind_connection_to_interface:
+            self.structured_config.ip_security.connection_tx_interface_match_source_ip = True
```
```suggestion
        if self.inputs.ipsec_settings.bind_connection_to_interface:
            self.structured_config.ip_security.connection_tx_interface_match_source_ip = True
```

gmuloc committed:
Apply suggestions from code review

Vibhu-gslab reviewed (approved):
LGTM!

gmuloc committed:
Fix: That's why you don't code in Github PR review UI

gmuloc commented on python-avd/pyavd/_eos_designs/structured_config/network_services/ip_security.py:
```diff
@@ -65,14 +65,6 @@ def ip_security(self: AvdStructuredConfigNetworkServicesProtocol) -> None:
-        Note: Duplicated in overlay/ip_security.py.
-        """
-        if self.inputs.ipsec_settings.bind_connection_to_interface:
+        if is_ipsec_policy_in_use:
```
```suggestion
        if is_ipsec_policy_in_use and self.inputs.ipsec_settings.bind_connection_to_interface:
```

gmuloc committed:
Update python-avd/pyavd/_eos_designs/structured_config/network_services/ip_security.py

gmuloc committed:
Merge branch 'devel' into eosDesignsBindIPSecConnSrcIntf

gmuloc committed:
Merge branch 'devel' into eosDesignsBindIPSecConnSrcIntf
